File: src/dispatch.ts
/**
 * CLI dispatch — parse top-level argv, route to subcommands.
 *
 * This module is the testable core of the CLI entry point. It accepts
 * injected I/O streams so tests can capture output without touching
 * process.stdout / process.stderr.
 */

import { ErrorCode, exitCodeFor, messageFor } from "./errors.ts";
import { formatErrorLine } from "./error-line.ts";
import { validateCommand } from "./validate.ts";
import { listCommand } from "./list.ts";
import { queryCommand } from "./query.ts";

// ---------------------------------------------------------------------------
// Public interface
// ---------------------------------------------------------------------------

export interface DispatchOptions {
  argv: string[];
  cwd: string;
  stdout: { write(s: string): void };
  stderr: { write(s: string): void };
  isTTY: boolean;
  version: string;
}

// ---------------------------------------------------------------------------
// Usage text
// ---------------------------------------------------------------------------

const USAGE = `Usage: yass <command> [options]

Commands:
  validate [paths...]   Validate .yass.yaml files
  query <name> [scope]  Query a spec by name
  list [path]           List discovered specs

Options:
  --help                Show this help message
  --version             Show version information
`;

// ---------------------------------------------------------------------------
// Known subcommands
// ---------------------------------------------------------------------------

const SUBCOMMANDS = ["validate", "query", "list"] as const;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Emit a single error line to stderr and return exit code 2.
 */
function usageError(
  stderr: { write(s: string): void },
  code: string,
  args: Record<string, string> = {},
): number {
  const msg = messageFor(code, args);
  const line = formatErrorLine({ code, message: msg });
  stderr.write(line + "\n");
  stderr.write(USAGE);
  return exitCodeFor(code);
}

/**
 * Check whether a token is an abbreviation of any known subcommand.
 * Returns true when the token is a proper prefix (length >= 1, length < full)
 * of at least one subcommand, and is NOT an exact match.
 */
function isAbbreviation(token: string): boolean {
  const lower = token.toLowerCase();
  for (const cmd of SUBCOMMANDS) {
    if (lower.length >= 1 && lower.length < cmd.length && cmd.startsWith(lower)) {
      return true;
    }
  }
  return false;
}

/**
 * Check whether a token is a case-mismatched form of a known subcommand.
 * Returns true when the lowered token exactly matches a subcommand but the
 * original token does not.
 */
function isCaseMismatch(token: string): boolean {
  const lower = token.toLowerCase();
  for (const cmd of SUBCOMMANDS) {
    if (lower === cmd && token !== cmd) {
      return true;
    }
  }
  return false;
}

/**
 * Known long flags in their canonical lowercase form.
 */
const KNOWN_FLAGS = ["--help", "--version"] as const;

/**
 * Check whether a long flag is a case-mismatched form of a known flag.
 * Returns true when the lowered flag exactly matches a known flag but the
 * original flag does not (e.g. "--Help", "--VERSION").
 */
function isFlagCaseMismatch(flag: string): boolean {
  const lower = flag.toLowerCase();
  for (const known of KNOWN_FLAGS) {
    if (lower === known && flag !== known) {
      return true;
    }
  }
  return false;
}

// ---------------------------------------------------------------------------
// Main dispatch
// ---------------------------------------------------------------------------

/**
 * Parse top-level argv, validate flags and subcommand, then route to the
 * appropriate handler. Returns a numeric exit code.
 */
export function dispatch(options: DispatchOptions): number {
  const { argv, cwd, stdout, stderr, isTTY, version } = options;

  // --- Scan for --help and --version anywhere in argv ---
  // --help takes priority over everything.
  // --version takes priority over errors but not --help.
  let hasHelp = false;
  let hasVersion = false;

  for (const arg of argv) {
    if (arg === "--help") hasHelp = true;
    if (arg === "--version") hasVersion = true;
  }

  if (hasHelp) {
    stdout.write(USAGE);
    return 0;
  }

  if (hasVersion) {
    stdout.write(`yass ${version}\n`);
    return 0;
  }

  // --- Walk argv to find the subcommand and validate flags ---
  // Flags (tokens starting with "-") are checked everywhere before "--".
  // The first non-flag, non-"--" token is the subcommand candidate.
  // Subsequent non-flag tokens are positional args for the subcommand.
  let subcommand: string | undefined;
  let doubleDashIndex = -1;

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]!;

    // End-of-options marker: stop flag checking.
    if (arg === "--") {
      doubleDashIndex = i;
      break;
    }

    // Empty argument
    if (arg === "") {
      return usageError(stderr, ErrorCode.ARGV_EMPTY_ARGUMENT);
    }

    // Bare "-" (stdin dash)
    if (arg === "-") {
      return usageError(stderr, ErrorCode.ARGV_STDIN_DASH);
    }

    // Long flag (not --help / --version, which were handled above)
    if (arg.startsWith("--")) {
      if (isFlagCaseMismatch(arg)) {
        return usageError(stderr, ErrorCode.ARGV_CASE_MISMATCH, { token: arg });
      }
      return usageError(stderr, ErrorCode.ARGV_UNKNOWN_FLAG, { flag: arg });
    }

    // Short flag
    if (arg.startsWith("-")) {
      return usageError(stderr, ErrorCode.ARGV_SHORT_FLAG, { flag: arg });
    }

    // Non-flag token: first one is the subcommand, rest are positional args.
    if (subcommand !== undefined) {
      // Positional arg for the subcommand — skip dispatch-level checks.
      continue;
    }

    // First non-flag token is the subcommand candidate.
    if ((SUBCOMMANDS as readonly string[]).includes(arg)) {
      subcommand = arg;
    } else if (isCaseMismatch(arg)) {
      return usageError(stderr, ErrorCode.ARGV_CASE_MISMATCH, { token: arg });
    } else if (isAbbreviation(arg)) {
      return usageError(stderr, ErrorCode.ARGV_ABBREVIATION, { token: arg });
    } else {
      return usageError(stderr, ErrorCode.ARGV_UNKNOWN_SUBCOMMAND, {
        arg,
      });
    }
  }

  // No subcommand found
  if (subcommand === undefined) {
    return usageError(stderr, ErrorCode.ARGV_NO_SUBCOMMAND);
  }

  // --- Build subcommand args ---
  // Collect positional args that follow the subcommand. The "--" token
  // (if present) is stripped; tokens after it are included verbatim.
  const subIdx = argv.indexOf(subcommand);
  const rawArgs = argv.slice(subIdx + 1);
  const subArgs = rawArgs.filter((a) => a !== "--");

  // --- Dispatch to subcommand ---
  switch (subcommand) {
    case "validate":
      return validateCommand(subArgs, cwd);

    case "list":
      return listCommand(subArgs, cwd, stdout, stderr, isTTY);

    case "query":
      return queryCommand(subArgs, cwd, stdout, stderr, isTTY);

    default:
      // Unreachable
      return usageError(stderr, ErrorCode.ARGV_UNKNOWN_SUBCOMMAND, {
        arg: subcommand,
      });
  }
}

File: src/dispatch.cpp
#include "dispatch.hpp"

#include <array>
#include <csignal>
#include <exception>
#include <optional>
#include <ostream>
#include <string>
#include <string_view>
#include <vector>

#include "diag.hpp"
#include "list.hpp"
#include "query.hpp"
#include "validate.hpp"

namespace yass::dispatch {

// --------------------------------------------------------------------------
// Top-level usage text — captured byte-for-byte from the reference `yass
// --help`. The reference reuses one fixed usage block; we reproduce it exactly
// so the differential happy-path (stdout + exit) holds.
// --------------------------------------------------------------------------
const char* const kUsageText =
    "Usage: yass <command> [arguments]\n"
    "\n"
    "Commands:\n"
    "  validate    Validate .yass.yaml files\n"
    "  list        List specs in .yass.yaml files\n"
    "  query       Query a spec by name\n"
    "\n"
    "Flags:\n"
    "  --help      Show this help\n"
    "  --version   Show version\n";

namespace {

// ASCII-only lowercase. Argv tokens are matched against the fixed ASCII
// keywords validate/query/list/--help/--version; an ASCII fold is correct and
// avoids locale dependence. Non-ASCII bytes pass through unchanged.
std::string ascii_lower(std::string_view s) {
    std::string out;
    out.reserve(s.size());
    for (char c : s) {
        if (c >= 'A' && c <= 'Z') c = static_cast<char>(c - 'A' + 'a');
        out.push_back(c);
    }
    return out;
}

// True iff `prefix` is a proper prefix of `whole` (strictly shorter, and `whole`
// begins with it).
bool is_proper_prefix(std::string_view prefix, std::string_view whole) {
    return prefix.size() < whole.size() && whole.substr(0, prefix.size()) == prefix;
}

// Emit one ErrorLine (no source file: <file> renders as the literal "yass") to
// `err` and return the spec-assigned exit code for `code`.
int emit_argv_error(std::ostream& err, diag::ErrorCode code, std::string_view arg) {
    diag::Diagnostic d;
    d.file = "";  // empty => the literal token "yass" per cli.ErrorLine
    d.code = code;
    d.message = diag::canonical_message(code, arg);
    err << diag::format_error_line(d) << '\n';
    return diag::exit_for(code);
}

// Canonical subcommands and global flags.
constexpr std::array<std::string_view, 3> kSubcommands = {"validate", "query", "list"};
constexpr std::array<std::string_view, 2> kFlags = {"--help", "--version"};

// Classify a flag-shaped token (it begins with '-', and is neither "-" nor
// "--"). Returns the argv error code for it. The reference rules, established
// empirically against the oracle:
//   - single dash (does not start with "--")        -> short_flag
//   - "--X" whose ASCII-lower equals a global flag   -> case_mismatch
//   - "--X" whose ASCII-lower is a proper prefix of a
//     global flag (and longer than "--")             -> abbreviation
//   - otherwise                                       -> unknown_flag
diag::ErrorCode classify_flag(std::string_view tok) {
    if (tok.size() < 2 || tok[1] != '-') {
        // Single-dash form: "-h", "-v", "-help", ...
        return diag::ErrorCode::argv_short_flag;
    }
    std::string lc = ascii_lower(tok);
    for (std::string_view f : kFlags) {
        if (lc == f) return diag::ErrorCode::argv_case_mismatch;
    }
    if (lc.size() > 2) {  // longer than the bare "--"
        for (std::string_view f : kFlags) {
            if (is_proper_prefix(lc, f)) return diag::ErrorCode::argv_abbreviation;
        }
    }
    return diag::ErrorCode::argv_unknown_flag;
}

// Classify the subcommand token (the first positional). Returns nullopt when it
// is an exact canonical subcommand (caller dispatches); otherwise the argv
// error code. Established empirically:
//   - exact canonical                                  -> dispatch (nullopt)
//   - ASCII-lower equals a canonical (wrong case)       -> case_mismatch
//   - ASCII-lower is a proper prefix of a canonical     -> abbreviation
//   - otherwise                                          -> unknown_subcommand
std::optional<diag::ErrorCode> classify_subcommand(std::string_view tok) {
    for (std::string_view s : kSubcommands) {
        if (tok == s) return std::nullopt;
    }
    std::string lc = ascii_lower(tok);
    for (std::string_view s : kSubcommands) {
        if (lc == s) return diag::ErrorCode::argv_case_mismatch;
    }
    for (std::string_view s : kSubcommands) {
        if (is_proper_prefix(lc, s)) return diag::ErrorCode::argv_abbreviation;
    }
    return diag::ErrorCode::argv_unknown_subcommand;
}

}  // namespace

int run_dispatch(const std::vector<std::string>& argv_tail, std::ostream& out, std::ostream& err,
                 int list_tty_width) {
    // --------------------------------------------------------------------
    // Phase 1 — global --help / --version scan.
    //
    // --help and --version are recognized ANYWHERE in argv, regardless of the
    // `--` end-of-options marker and regardless of any argv error that would
    // otherwise fire. --help wins over --version when both appear. Only the
    // EXACT lowercase spellings count here; wrong-case / abbreviated spellings
    // fall through to the argv-error classification below.
    // --------------------------------------------------------------------
    bool saw_help = false;
    bool saw_version = false;
    for (const std::string& tok : argv_tail) {
        if (tok == "--help") saw_help = true;
        else if (tok == "--version") saw_version = true;
    }
    if (saw_help) {
        out << kUsageText;
        return diag::ExitCode::SUCCESS;
    }
    if (saw_version) {
        out << "yass " << YASS_VERSION << '\n';
        return diag::ExitCode::SUCCESS;
    }

    // --------------------------------------------------------------------
    // Locate the single global `--` end-of-options marker (the FIRST "--"
    // token). Tokens at and after it are no longer flag-classified; the marker
    // itself is consumed and not passed to any handler. Empty-string and bare
    // "-" tokens remain argv errors even past the marker.
    // --------------------------------------------------------------------
    std::optional<std::size_t> dd_idx;
    for (std::size_t i = 0; i < argv_tail.size(); ++i) {
        if (argv_tail[i] == "--") {
            dd_idx = i;
            break;
        }
    }
    auto past_marker = [&](std::size_t i) { return dd_idx.has_value() && i > *dd_idx; };

    // --------------------------------------------------------------------
    // Phase 2 — token-level argv errors, left to right; the FIRST offending
    // token wins. These take precedence over the subcommand classification
    // (e.g. `bogus -h` reports short_flag, not unknown_subcommand).
    //   - empty string            -> empty_argument (even past the marker)
    //   - bare "-"                -> stdin_dash      (even past the marker)
    //   - flag-shaped, pre-marker -> classify_flag (short/case/abbrev/unknown)
    // --------------------------------------------------------------------
    for (std::size_t i = 0; i < argv_tail.size(); ++i) {
        if (dd_idx && i == *dd_idx) continue;  // the marker is not classified
        const std::string& tok = argv_tail[i];
        if (tok.empty()) {
            return emit_argv_error(err, diag::ErrorCode::argv_empty_argument, tok);
        }
        if (tok == "-") {
            return emit_argv_error(err, diag::ErrorCode::argv_stdin_dash, tok);
        }
        if (!past_marker(i) && tok[0] == '-' && tok != "--") {
            return emit_argv_error(err, classify_flag(tok), tok);
        }
    }

    // --------------------------------------------------------------------
    // Phase 3 — subcommand selection. Positionals are all tokens except the
    // consumed `--` marker. The first positional is the subcommand; the rest
    // are the handler's arguments (verbatim, marker already removed).
    // --------------------------------------------------------------------
    std::vector<std::string> positionals;
    positionals.reserve(argv_tail.size());
    for (std::size_t i = 0; i < argv_tail.size(); ++i) {
        if (dd_idx && i == *dd_idx) continue;
        positionals.push_back(argv_tail[i]);
    }

    if (positionals.empty()) {
        return emit_argv_error(err, diag::ErrorCode::argv_no_subcommand, {});
    }

    const std::string& sub = positionals.front();
    std::vector<std::string> handler_args(positionals.begin() + 1, positionals.end());

    std::optional<diag::ErrorCode> sub_err = classify_subcommand(sub);
    if (sub_err) {
        return emit_argv_error(err, *sub_err, sub);
    }

    // --------------------------------------------------------------------
    // Phase 4 — dispatch. Any uncaught exception from a handler becomes one
    // yass.internal.uncaught ErrorLine and exit 1 (cli.Dispatch INVARIANT).
    // --------------------------------------------------------------------
    try {
        if (sub == "validate") return yass::validate::run_validate(handler_args, out, err);
        if (sub == "query") return yass::query::run_query(handler_args, out, err);
        // sub == "list"
        return yass::list::run_list(handler_args, out, err, list_tty_width);
    } catch (const std::exception& e) {
        return emit_argv_error(err, diag::ErrorCode::internal_uncaught, e.what());
    } catch (...) {
        return emit_argv_error(err, diag::ErrorCode::internal_uncaught, "unknown error");
    }
}

int signal_exit_code(int sig) {
    switch (sig) {
        case SIGINT:
            return diag::ExitCode::ON_SIGINT;  // 130
        case SIGTERM:
            return diag::ExitCode::ON_SIGTERM;  // 143
        default:
            return diag::ExitCode::SUCCESS;
    }
}

}  // namespace yass::dispatch
